SSL support for libpq

`../configure $MY_STD_CONFIGURE_FLAGS --without-readline --with-openssl`
`checking for CRYPTO_new_ex_data in -lcrypto... no`
`configure: error: library 'crypto' is required for OpenSSL`

HINTS:
http://stackoverflow.com/questions/2952572/building-openssl-on-android-ndk
http://pgolub.wordpress.com/2008/12/15/building-postgresql-client-library-using-mingw-under-winxp-sp3/
https://github.com/fries/android-external-openssl/

More Hints:
https://www.nixuopen.org/media/library/files/2011/05/racoon_for_android_compile_problems_and_resolutions.html
https://groups.google.com/forum/#!topic/android-ndk/DMftXJOCHsQ
http://android3.git.kernel.org/platform/external/openssl.git/
https://github.com/fries/android-external-openssl/blob/master/README.android
https://groups.google.com/forum/#!msg/android-ndk/dM70p9b-yBk/C2ZMGE7aqd0J

`D/dalvikvm( 1972): Trying to load lib /data/data/org.qgis.qgis/lib/libpq.so 0x40791a90
W/System.err( 1972): java.lang.reflect.InvocationTargetException
W/System.err( 1972):     at java.lang.reflect.Method.invokeNative(Native Method)
W/System.err( 1972):     at java.lang.reflect.Method.invoke(Method.java:491)
W/System.err( 1972):     at org.kde.necessitas.origo.QtActivity.loadApplication(QtActivity.java:144)
W/System.err( 1972):     at org.kde.necessitas.origo.QtActivity.access$200(QtActivity.java:77)
W/System.err( 1972):     at org.kde.necessitas.origo.QtActivity$3$1$1.run(QtActivity.java:201)
W/System.err( 1972):     at android.os.Handler.handleCallback(Handler.java:587)
W/System.err( 1972):     at android.os.Handler.dispatchMessage(Handler.java:92)
W/System.err( 1972):     at android.os.Looper.loop(Looper.java:132)
W/System.err( 1972):     at android.app.ActivityThread.main(ActivityThread.java:4123)
W/System.err( 1972):     at java.lang.reflect.Method.invokeNative(Native Method)
W/System.err( 1972):     at java.lang.reflect.Method.invoke(Method.java:491)
W/System.err( 1972):     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:841)
W/System.err( 1972):     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:599)
W/System.err( 1972):     at dalvik.system.NativeStart.main(Native Method)
W/System.err( 1972): Caused by: java.lang.UnsatisfiedLinkError: Cannot load library: reloc_library[1315]:    87 cannot locate 'ENGINE_load_private_key'...
W/System.err( 1972):     at java.lang.Runtime.load(Runtime.java:390)
W/System.err( 1972):     at java.lang.System.load(System.java:534)
W/System.err( 1972):     at org.kde.necessitas.industrius.QtNative.loadBundledLibraries(QtNative.java:116)
W/System.err( 1972):     at org.kde.necessitas.industrius.QtActivityDelegate.loadApplication(QtActivityDelegate.java:156)
W/System.err( 1972):     ... 14 more`

--- a/distribute.sh
+++ b/distribute.sh
@@ -116,7 +116,7 @@ fi
 
 case $OSTYPE in
   darwin*)
-    SED="sed -i ''"
+    SED="sed -i .orig"
     ;;
   *)
     SED="sed -i"

--- a/recipes/freexl/recipe.sh
+++ b/recipes/freexl/recipe.sh
@@ -47,10 +47,11 @@ function build_freexl() {
   try mkdir -p $BUILD_PATH/freexl/build-$ARCH
   try cd $BUILD_PATH/freexl/build-$ARCH
   push_arm
+
   export LDFLAGS="$LDFLAGS -liconv"
-  error $LDFLAGS
   try $BUILD_freexl/configure --prefix=$STAGE_PATH --host=${TOOLCHAIN_PREFIX}
   try $MAKESMP install
+
   pop_arm
 }
 

--- a/recipes/iconv/recipe.sh
+++ b/recipes/iconv/recipe.sh
@@ -48,11 +48,14 @@ function shouldbuild_iconv() {
 function build_iconv() {
   try mkdir -p $BUILD_PATH/iconv/build-$ARCH
   try cd $BUILD_PATH/iconv/build-$ARCH
-	push_arm
+
+  push_arm
+
   try $BUILD_iconv/configure --prefix=$STAGE_PATH --host=${TOOLCHAIN_PREFIX}
   try $MAKESMP
   try make install
-	pop_arm
+
+  pop_arm
 }
 
 # function called after all the compile have been done

--- a/recipes/libzip/recipe.sh
+++ b/recipes/libzip/recipe.sh
@@ -31,7 +31,7 @@ function prebuild_libzip() {
 
 function shouldbuild_libzip() {
   # If lib is newer than the sourcecode skip build
-  if [ $BUILD_PATH/libzip/build-$ARCH/src/.libs/liblibzip.so -nt $BUILD_libzip/.patched ]; then
+  if [ $BUILD_PATH/libzip/master/libs/$ARCH/libzip.so -nt $BUILD_libzip/.patched ]; then
     DO_BUILD=0
   fi
 }

--- a/recipes/openssl/recipe.sh
+++ b/recipes/openssl/recipe.sh
@@ -31,6 +31,9 @@ function prebuild_openssl() {
   echo "patching $BUILD_PATH/openssl/openssl-${VERSION_openssl}/config"
   try $SED 's/armv\[7-9\]\*-\*-android/armeabi-v7a\*-\*-android|armv\[7-9\]\*-\*-android/g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/config
 
+  echo "patching $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Configure"
+  LC_ALL=C try $SED 's/SHLIB_EXT=$shared_extension/SHLIB_EXT=.so/g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Configure
+
   touch .patched
 }
 
@@ -75,6 +78,16 @@ function build_openssl() {
       -D__ANDROID_API__=$ANDROIDAPI \
       --prefix=/  \
       $(eval echo $CFLAGS_WITHOUT_SYSROOT)
+
+  echo "patching $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile"
+  # remove docs
+  try $SED '646,690d' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile
+  try $SED '621,643d' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile
+  # remove "link-shared" target, since we do not want links
+  try $SED '346,352d' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile
+  # remove so.x.y versions
+  try $SED 's/LIBVERSION=$(SHLIB_MAJOR).$(SHLIB_MINOR)//g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile
+  try $SED 's/LIBCOMPATVERSIONS=";$(SHLIB_VERSION_HISTORY)"//g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/Makefile
 
   ${MAKESMP} depend
   ${MAKESMP} CALC_VERSIONS="SHLIB_COMPAT=; SHLIB_SOVER=" build_libs

--- a/recipes/openssl/recipe.sh
+++ b/recipes/openssl/recipe.sh
@@ -36,7 +36,7 @@ function prebuild_openssl() {
 
 function shouldbuild_openssl() {
   # If lib is newer than the sourcecode skip build
-  if [ $BUILD_openssl/build-$ARCH/libssl.so -nt $BUILD_openssl/.patched ]; then
+  if [ $BUILD_openssl/libssl.so -nt $BUILD_openssl/.patched ]; then
     DO_BUILD=0
   fi
 }

--- a/recipes/openssl/recipe.sh
+++ b/recipes/openssl/recipe.sh
@@ -1,11 +1,10 @@
 #!/bin/bash
 
-# NOTES
-# 1) The SYS_getrandom is disabled by patch, look at openssl FAQ about random function
-# 2) Built without threading support
- 
+# OpenSSL 1.1.x has new API, so ATM with the new API it is not possible to build QCA
+# To see how to build OpenSSL 1.1.x, check  https://github.com/opengisch/OSGeo4A/pull/25
+
 # version of your package
-VERSION_openssl=1.1.1
+VERSION_openssl=1.0.2o
 
 # dependencies of this recipe
 DEPS_openssl=()
@@ -23,11 +22,14 @@ BUILD_openssl=$BUILD_PATH/openssl/$(get_directory $URL_openssl)
 # function called for preparing source code if needed
 # (you can apply patch etc here.)
 function prebuild_openssl() {
+  cd $BUILD_PATH/openssl/openssl-${VERSION_openssl}
+  # check marker
+  if [ -f .patched ]; then
+    return
+  fi
+
   echo "patching $BUILD_PATH/openssl/openssl-${VERSION_openssl}/config"
-  try sed -i 's/android-armeabi/android-arm/g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/config
-  
-  # ad random seed & patch of rand_unix.c: https://mta.openssl.org/pipermail/openssl-users/2018-September/008860.html 
-  try patch -p1 $BUILD_PATH/openssl/openssl-${VERSION_openssl}/crypto/rand/rand_unix.c $RECIPE_openssl/patches/rand_unix.patch
+  try $SED 's/armv\[7-9\]\*-\*-android/armeabi-v7a\*-\*-android|armv\[7-9\]\*-\*-android/g' $BUILD_PATH/openssl/openssl-${VERSION_openssl}/config
 
   touch .patched
 }
@@ -41,8 +43,9 @@ function shouldbuild_openssl() {
 
 # function called to build the source code
 function build_openssl() {
-  try mkdir -p $BUILD_PATH/openssl/build-$ARCH
-  try cd $BUILD_PATH/openssl/build-$ARCH
+  # unfortunately config and Configure uses relative paths to this
+  # se we need to do in-source build
+  try cd $BUILD_openssl
 
   push_arm
 
@@ -59,33 +62,25 @@ function build_openssl() {
   export SYSTEM=android
   # export ARCH is exported in distribute.sh
 
-  export CROSS_COMPILE=$ANDROID_EABI-
-  export ANDROID_DEV=$_ANDROID_NDK/platforms/android-$_ANDROID_API/$_ANDROID_ARCH/usr
+  export ANDROID_DEV=$_ANDROID_NDK/platforms/$_ANDROID_API/$_ANDROID_ARCH/usr
   export HOSTCC=gcc
-
   export ANDROID_TOOLCHAIN=$ANDROIDNDK/toolchains/$TOOLCHAIN_BASEDIR-$TOOLCHAIN_VERSION/prebuilt/$PYPLATFORM-x86_64/bin
-     
-  # tools are prefixed in config
-  export CC=$NDK_TOOLCHAIN_BASENAMEgcc
-  export AR=$NDK_TOOLCHAIN_BASENAMEar
-  export CXX=$NDK_TOOLCHAIN_BASENAMEg++
-  export LINK=${CXX}
-  export LD=$NDK_TOOLCHAIN_BASENAMEld
-  export RANLIB=$NDK_TOOLCHAIN_BASENAMEranlib
-  export STRIP=$NDK_TOOLCHAIN_BASENAMEstrip
-  
-  # sysroot confuses the config command
+
   CFLAGS_WITHOUT_QUOTES=$(eval echo $CFLAGS)
   CFLAGS_WITHOUT_SYSROOT=${CFLAGS_WITHOUT_QUOTES//"--sysroot $NDKPLATFORM"/}
 
-  # no-threads: build falling on ./libcrypto.so: error: undefined reference to 'pthread_atfork'
-  ./../openssl-${VERSION_openssl}/config no-threads shared no-comp no-hw no-engine --openssldir=/usr/local/ssl/$ANDROIDAPI/ -D__ANDROID_API__=$ANDROIDAPI --prefix=/  $(eval echo $CFLAGS_WITHOUT_SYSROOT)
-    
+  ./config \
+      shared no-hw \
+      --openssldir=/usr/local/ssl/$ANDROIDAPI/ \
+      -D__ANDROID_API__=$ANDROIDAPI \
+      --prefix=/  \
+      $(eval echo $CFLAGS_WITHOUT_SYSROOT)
+
   ${MAKESMP} depend
   ${MAKESMP} CALC_VERSIONS="SHLIB_COMPAT=; SHLIB_SOVER=" build_libs
 
   # install
-  try ${MAKE} DESTDIR=$STAGE_PATH install
+  try ${MAKE} INSTALL_PREFIX=$STAGE_PATH install
     
   pop_arm
 }

--- a/recipes/postgresql/recipe.sh
+++ b/recipes/postgresql/recipe.sh
@@ -4,7 +4,7 @@
 VERSION_postgresql=10.1
 
 # dependencies of this recipe
-DEPS_postgresql=(iconv)
+DEPS_postgresql=(iconv openssl)
 
 # url of the package
 URL_postgresql=https://ftp.postgresql.org/pub/source/v${VERSION_postgresql}/postgresql-${VERSION_postgresql}.tar.bz2
@@ -51,7 +51,7 @@ function build_postgresql() {
   LIBS="-lgnustl_shared -lsupc++ -lstdc++" \
   LDFLAGS="${LDFLAGS} -L$ANDROIDNDK/sources/cxx-stl/gnu-libstdc++/$TOOLCHAIN_VERSION/libs/${ARCH}" \
   USE_DEV_URANDOM=1 \
-  try $BUILD_postgresql/configure --prefix=$STAGE_PATH --host=${TOOLCHAIN_PREFIX} --without-readline
+  try $BUILD_postgresql/configure --prefix=$STAGE_PATH --host=${TOOLCHAIN_PREFIX} --without-readline --with-openssl
   try $MAKESMP -C src/interfaces/libpq
 
   #simulate make install

--- a/recipes/qca/recipe.sh
+++ b/recipes/qca/recipe.sh
@@ -39,7 +39,7 @@ function prebuild_qca() {
 
 function shouldbuild_qca() {
  # If lib is newer than the sourcecode skip build
- if [ $BUILD_qca/build-$ARCH/libqca2.so -nt $BUILD_qca/.patched ]; then
+ if [ $BUILD_qca/build-$ARCH/lib/libqca-qt5.so -nt $BUILD_qca/.patched ]; then
   DO_BUILD=0
  fi
 }

--- a/recipes/qca/recipe.sh
+++ b/recipes/qca/recipe.sh
@@ -4,7 +4,7 @@
 VERSION_qca=2.1.3
 
 # dependencies of this recipe
-DEPS_qca=()
+DEPS_qca=(openssl)
 
 # url of the package
 # URL_qca=http://delta.affinix.com/download/qca/2.0/qca-${VERSION_qca}.tar.gz

--- a/recipes/qtkeychain/recipe.sh
+++ b/recipes/qtkeychain/recipe.sh
@@ -37,7 +37,7 @@ function prebuild_qtkeychain() {
 
 function shouldbuild_qtkeychain() {
  # If lib is newer than the sourcecode skip build
- if [ $BUILD_qtkeychain/build-$ARCH/libqtkeychain2.so -nt $BUILD_qtkeychain/.patched ]; then
+ if [ $BUILD_qtkeychain/build-$ARCH/libqt5keychain.so -nt $BUILD_qtkeychain/.patched ]; then
   DO_BUILD=0
  fi
 }

